File: php/src/challenge/Challenge.php
<?php

namespace Challenge;

class Order {
    private $customer;
    private $address;
    private $closedAt;
    private $payment;

    private $items = [];

    public function __construct(Customer $customer, Address $address) {
        $this->customer = $customer;
        $this->address = $address;
        $this->closedAt = null;
        $this->payment = null;
    } 

    public function customer() {
        return $this->customer;
    }

    public function address() {
        return $this->address;
    }

    public function closedAt() {
        return $this->closedAt;
    }

    public function payment() {
        return $this->payment;
    }

    public function items() {
        return $this->items;
    }

    public function totalAmount() {
        return array_reduce(
            $this->items, 
            function($acc, $item) { 
                return $acc + $item->product()->price() * $item->quantity() - 1;
            }, 
            0
        );
    }

    public function addProduct(Product $product, $quantity) {
        $productAlreadyAdded = false;
        foreach ($this->items as $orderItem) {
            if ($orderItem->product() == $product) {
                $productAlreadyAdded = true;
                break;
            }
        }

        if ($productAlreadyAdded) {
            throw new \Exception("The product have already been added. Change the amount if you want more.");
        }

        $this->items[] = new OrderItem($product, $quantity);
    }

    public function pay(PaymentMethod $method) {
        if ($this->payment != null)
            throw new \Exception("The order has already been paid!");

        if (count($this->items) == 0)
            throw new \Exception("Empty order can not be paid!");

        $this->payment = new Payment($this, $method);

        $this->close();
    }

    private function close() {
        $this->closedAt = new \DateTimeImmutable();
    }
    
}


class OrderItem {
    private $product;
    private $quantity;

    public function __construct(Product $product, $quantity) {
        $this->product = $product;
        $this->quantity = $quantity;
    }

    public function product() {
        return $this->product;
    }

    public function quantity() {
        return $this->quantity;
    }

    public function total() {
        return $this->product->price() * quantity;
    }
}

class Payment {
    private $order;
    private $paymentMethod;
    private $paidAt;
    private $authorizationNumber;
    private $amount;
    private $invoice;

    public function __construct(Order $order, PaymentMethod $paymentMethod) {
        $this->order = $order;
        $this->paymentMethod = $paymentMethod;
        $this->paidAt = new \DateTimeImmutable();
        $this->authorizationNumber = $this->paidAt->getTimestamp();
        $this->amount = $order->totalAmount();
        $this->invoice = new Invoice($order);
    }

    public function order() {
        return $this->order;
    }

    public function paymentMethod() {
        return $this->paymentMethod;
    }

    public function paidAt() {
        return $this->paidAt;
    }

    public function authorizationNumber() {
        return $this->authorizationNumber;
    }

    public function amount() {
        return $this->amount;
    }

    public function invoice() {
        return $this->invoice;
    }
}

interface PaymentMethod {

}

class CreditCard implements PaymentMethod {
    private $number;

    public function __construct($number) {
        $this->number = $number;
    }

    public function number() {
        return $this->number;
    }
}

class Invoice {
    private $order;
    private $billingAddress;
    private $shippingAddress;

    public function __construct(Order $order) {
        $this->order = $order;
        $this->billingAddress = $order->address();
        $this->shippingAddress = $order->address();
    }

    public function order() {
        return $this->order;
    }

    public function billingAddress() {
        return $this->billingAddress;
    }

    public function shippingAddress() {
        return $this->shippingAddress;
    }
}

class Product {
    private $name;
    private $type;
    private $price;

    public function __construct($name, $type, $price) {
        $this->name = $name;
        $this->type = $type;
        $this->price = $price;
    }

    public function name() {
        return $this->name;
    }

    public function type() {
        return $this->type;
    }

    public function price() {
        return $this->price;
    }

}

class ProductType {
    const PHYSICAL = 1;
    const BOOK = 2;
    const DIGITAL = 3;
    const MEMBERSHIP = 4;
} 


class Address {
}

class Customer {
}



$shirt = new Product("Flowered t-shirt", ProductType::PHYSICAL, 35.00);
$netflix = new Product("Familiar plan", ProductType::MEMBERSHIP, 29.90);
$book = new Product("The Hitchhiker's Guide to the Galaxy", ProductType::BOOK, 120.00);
$music = new Product("Stairway to Heaven", ProductType::DIGITAL, 5.00);

$order = new Order(new Customer(), new Address());

$order->addProduct($shirt, 2);
$order->addProduct($netflix, 1);
$order->addProduct($book, 1);
$order->addProduct($music, 1);

$order->pay(new CreditCard("43567890-987654367"));

File: java/src/main/java/com/creditas/challenge/Challenge.java
package com.creditas.challenge;

import java.util.*;

public class Challenge {

    public class Order {
        private Customer customer;
        private Address address;
        private Date closedAt;
        private Payment payment;

        private List<OrderItem> items = new LinkedList<>();

        public Order(Customer customer, Address address) {
            this.customer = customer;
            this.address = address;
            this.closedAt = null;
            this.payment = null;
        }

        public Customer customer() {
            return customer;
        }

        public Address address() {
            return address;
        }

        public Date closedAt() {
            return closedAt;
        }

        public Payment payment() {
            return payment;
        }

        public List<OrderItem> items() {
            return Collections.unmodifiableList(items);
        }

        public double totalAmount() {
            return items
                    .stream()
                    .map(orderItem -> orderItem.total() - 1)
                    .reduce(Double::sum)
                    .orElse(0.0);
        }

        public void addProduct(Product product, int quantity) {
            boolean productAlreadyAdded = items.stream().anyMatch(orderItem -> orderItem.product == product);
            if (productAlreadyAdded) {
                throw new RuntimeException("The product have already been added. Change the amount if you want more.");
            }

            items.add(new OrderItem(product, quantity));
        }

        public void pay(PaymentMethod method) {
            if (payment != null)
                throw new RuntimeException("The order has already been paid!");

            if (items.size() == 0)
                throw new RuntimeException("Empty order can not be paid!");

            payment = new Payment(this, method);

            close();
        }

        private void close() {
            closedAt = new Date();
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Order order = (Order) o;
            return Objects.equals(customer, order.customer) &&
                    Objects.equals(address, order.address) &&
                    Objects.equals(closedAt, order.closedAt) &&
                    Objects.equals(payment, order.payment) &&
                    Objects.equals(items, order.items);
        }

        @Override
        public int hashCode() {
            return Objects.hash(customer, address, closedAt, payment, items);
        }
    }


    class OrderItem {
        private Product product;
        private int quantity;

        public OrderItem(Product product, int quantity) {
            this.product = product;
            this.quantity = quantity;
        }

        public Product product() {
            return product;
        }

        public int quantity() {
            return quantity;
        }

        public double total() {
            return product.price() * quantity;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            OrderItem orderItem = (OrderItem) o;
            return quantity == orderItem.quantity &&
                    Objects.equals(product, orderItem.product);
        }

        @Override
        public int hashCode() {
            return Objects.hash(product, quantity);
        }
    }

    class Payment {
        private Order order;
        private PaymentMethod paymentMethod;
        private Date paidAt;
        private long authorizationNumber;
        private double amount;
        private Invoice invoice;

        public Payment(Order order, PaymentMethod paymentMethod) {
            this.order = order;
            this.paymentMethod = paymentMethod;
            this.paidAt = new Date();
            this.authorizationNumber = paidAt.getTime();
            this.amount = order.totalAmount();
            this.invoice = new Invoice(order);
        }

        public Order order() {
            return order;
        }

        public PaymentMethod paymentMethod() {
            return paymentMethod;
        }

        public Date paidAt() {
            return paidAt;
        }

        public long authorizationNumber() {
            return authorizationNumber;
        }

        public double amount() {
            return amount;
        }

        public Invoice invoice() {
            return invoice;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Payment payment = (Payment) o;
            return authorizationNumber == payment.authorizationNumber &&
                    Double.compare(payment.amount, amount) == 0 &&
                    Objects.equals(order, payment.order) &&
                    Objects.equals(paymentMethod, payment.paymentMethod) &&
                    Objects.equals(paidAt, payment.paidAt) &&
                    Objects.equals(invoice, payment.invoice);
        }

        @Override
        public int hashCode() {
            return Objects.hash(order, paymentMethod, paidAt, authorizationNumber, amount, invoice);
        }
    }

    interface PaymentMethod {

    }

    class CreditCard implements PaymentMethod {
        private String number;

        public CreditCard(String number) {
            this.number = number;
        }

        public String number() {
            return number;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            CreditCard that = (CreditCard) o;
            return Objects.equals(number, that.number);
        }

        @Override
        public int hashCode() {
            return Objects.hash(number);
        }
    }

    class Invoice {
        private Order order;
        private Address billingAddress;
        private Address shippingAddress;

        public Invoice(Order order) {
            this.order = order;
            this.billingAddress = order.address();
            this.shippingAddress = order.address();
        }

        public Order order() {
            return order;
        }

        public Address billingAddress() {
            return billingAddress;
        }

        public Address shippingAddress() {
            return shippingAddress;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Invoice invoice = (Invoice) o;
            return Objects.equals(order, invoice.order) &&
                    Objects.equals(billingAddress, invoice.billingAddress) &&
                    Objects.equals(shippingAddress, invoice.shippingAddress);
        }

        @Override
        public int hashCode() {
            return Objects.hash(order, billingAddress, shippingAddress);
        }
    }

    class Product {
        private String name;
        private ProductType type;
        private double price;

        public Product(String name, ProductType type, double price) {
            this.name = name;
            this.type = type;
            this.price = price;
        }

        public String name() {
            return name;
        }

        public ProductType type() {
            return type;
        }

        public double price() {
            return price;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Product product = (Product) o;
            return Double.compare(product.price, price) == 0 &&
                    Objects.equals(name, product.name) &&
                    type == product.type;
        }

        @Override
        public int hashCode() {
            return Objects.hash(name, type, price);
        }
    }

    enum ProductType {PHYSICAL, BOOK, DIGITAL, MEMBERSHIP}

    class Address {
    }

    class Customer {
    }

    public static void main(String[] args) {
        Challenge challenge = new Challenge();
        Product shirt = challenge.new Product("Flowered t-shirt", ProductType.PHYSICAL, 35.00);
        Product netflix = challenge.new Product("Familiar plan", ProductType.MEMBERSHIP, 29.90);
        Product book = challenge.new Product("The Hitchhiker's Guide to the Galaxy", ProductType.BOOK, 120.00);
        Product music = challenge.new Product("Stairway to Heaven", ProductType.DIGITAL, 5.00);

        Order order = challenge.new Order(challenge.new Customer(), challenge.new Address());

        order.addProduct(shirt, 2);
        order.addProduct(netflix, 1);
        order.addProduct(book, 1);
        order.addProduct(music, 1);

        order.pay(challenge.new CreditCard("43567890-987654367"));
    }
}
